rfd/0157-aws-eks-discover.md

```diff
@@ -0,0 +1,225 @@
+---
+authors:  Anton Miniailo (anton@goteleport.com)
+state: Draft
+---
+
+# RFD 0157 - AWS EKS Discover Integration
+
+## Required Approvals
+
+* Engineering: @r0mant && (@tigrato || @marcoandredinis)
+* Product: @xinding33 || @klizhentas
+* Security: @reedloden || @jentfoo
+
+## What
+
+Add a guided workflow to the Discover UI for enrolling AWS Elastic Kubernetes Service (EKS) clusters.
+
+## Why
+
+At present, the process for adding an EKS cluster to Teleport is somewhat technical. Users need to either manually install the Teleport agent on the cluster or manually set up and run discovery and kube services for EKS.
+We aim to lower the friction of adding EKS clusters to Teleport by providing a seamless experience. Creating a guided UI workflow will help users onboard
+their EKS clusters more quickly and with less effort.
+
+## Scope
+
+This RFD focuses on Amazon EKS clusters. A similar approach can be taken later for other specialized Kubernetes providers, such as
+Azure AKS and Google GKE.
+
+## Details
+
+### UX
+
+We will build a workflow similar to the ones we already have for AWS RDS and EC2 enrollment.
+
+![EKS screen](assets/0157-eks-screen.png)
+
+General steps of the workflow:
+
+1. AWS Integration setup
+2. Clusters Enrollment
+3. Setup access
+4. Test connection
+
+AWS Integration for EKS will require permissions to list/describe EKS clusters and associate an identity provider
+to an EKS cluster (details [below](#eks-identity-provider-association)). AWS permissions config:
+
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "eks:DescribeCluster",
+                "eks:ListClusters",
+                "eks:AssociateIdentityProviderConfig",
+                "eks:DisassociateIdentityProviderConfig",
+                "eks:ListIdentityProviderConfigs",
+                "eks:DescribeIdentityProviderConfig"
+            ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+
+We will reuse existing code for AWS Integration setup using these new permissions.
+
+After setting up AWS integration, users will see a list of the EKS clusters in the next step.
+There, they will be able to select the cluster they want to enroll.
+We will determine which clusters are already enrolled and present in the Teleport inventory; those clusters
+will be greyed out in the table.
+
+Users will also have an option to enable Kubernetes App Discovery in the enrolled clusters, this option
+will be enabled by default.
+
+Users will have two main options for enrolling EKS clusters:
+ - Enroll specific selected cluster from the list of available ones.
+ - Set up automatic discovery and enrollment of the clusters using OIDC integration.
+
+Cluster enrollment will be done through the installation of the Teleport agent using the Helm chart
+`teleport-kube-agent`.
+
+### Enrolling Selected Clusters
```

Comment:
We should reiterate that we won't be able to support EKS private clusters where the API is not "publicly" exposed to the Teleport service.
EKS apis can be private and run under restrictive VPC rules without accepting incoming traffic from the internet. For those clusters, we either force discovery service to run within the same VPC or we should just leave it out of scope.